## Derivatives

Getting intuitive understanding of derivatives by exploring a simple way of calculating derivatives arithmetically, also exploring a handy Python library that allows us to calculate derivatives symbolically.

In [1]:
# Importing the sympy library
from sympy import symbols, diff

## Informal definition of derivatives

The derivative of a function describes how the output of a function changes when there is a small change in an input variable. Let's use the cost function 𝐽(𝑤) as an example. The cost 𝐽 is the output and 𝑤 is the input variable. Let's give a 'small change' a name epsilon or 𝜖 (0.001)

This just says if you change the input to the function 𝐽(𝑤) by a little bit and the output changes by 𝑘 times that little bit, then the derivative of  𝐽(𝑤) is equal to 𝑘.

In [2]:
# Let's look at the derivative of the function 𝐽(𝑤)=𝑤2 at the point 𝑤=3 and 𝜖=0.001
J = (3)**2
J_epsilon = (3 + 0.001)**2
k = (J_epsilon - J)/0.001    # difference divided by epsilon
print(f"J = {J}, J_epsilon = {J_epsilon}, dJ_dw ~= k = {k:0.6f} ")

J = 9, J_epsilon = 9.006001, dJ_dw ~= k = 6.001000 


We have increased the input value a little bit (0.001), causing the output to change from 9 to 9.006001, an increase of 6 times the input increase. Our calculation above is not exactly 6 because to be exactly correct 𝜖 would need to be infinitesimally small or really, really small. That is why we use the symbols ≈ or ~= rather than =

In [3]:
# Let's see what happens if we make 𝜖 smaller.
J = (3)**2
J_epsilon = (3 + 0.000000001)**2
k = (J_epsilon - J)/0.000000001
print(f"J = {J}, J_epsilon = {J_epsilon}, dJ_dw ~= k = {k} ")

J = 9, J_epsilon = 9.000000006, dJ_dw ~= k = 6.000000496442226 


The value gets close to exactly 6 as we reduce the size of 𝜖.

## Finding symbolic derivatives

In backprop it is useful to know the derivative of simple functions at any input value. Put another way, we would like to know the 'symbolic' derivative rather than the 'arithmetic' derivative. With the symbolic derivative we can find the value of the derivative at any input value 𝑤.

𝐽=𝑤2
 
Defining the python variables and their symbolic names.

In [4]:
J, w = symbols('J, w')

Defining and printing the expression. SymPy produces a latex string which generates a readable equation.

In [5]:
J=w**2
J

w**2

Using SymPy's diff to differentiate the expression for 𝐽 with respect to 𝑤.

In [6]:
dJ_dw = diff(J,w)
dJ_dw

2*w

Evaluating the derivative at a few points by 'substituting' numeric values for the symbolic values. In the first example, 𝑤 is replaced by 2.

In [7]:
dJ_dw.subs([(w,2)])    # derivative at the point w = 2

4

In [8]:
dJ_dw.subs([(w,3)])    # derivative at the point w = 3

6

In [9]:
dJ_dw.subs([(w,-3)])    # derivative at the point w = -3

-6

𝐽=2𝑤

In [11]:
w, J = symbols('w, J')

In [12]:
J = 2 * w
J

2*w

In [13]:
dJ_dw = diff(J,w)
dJ_dw

2

In [14]:
dJ_dw.subs([(w,-3)])    # derivative at the point w = -3

2

Comparing this with the arithmetic calculation

In [15]:
J = 2*3
J_epsilon = 2*(3 + 0.001)
k = (J_epsilon - J)/0.001
print(f"J = {J}, J_epsilon = {J_epsilon}, dJ_dw ~= k = {k} ")

J = 6, J_epsilon = 6.002, dJ_dw ~= k = 1.9999999999997797 


For the function 𝐽=2𝑤, it is easy to see that any change in 𝑤 will result in 2 times that amount of change in the output 𝐽, regardless of the starting value of 𝑤. Our NumPy and arithmetic results confirm this.

𝐽=𝑤3

In [16]:
J, w = symbols('J, w')

In [17]:
J=w**3
J

w**3

In [18]:
dJ_dw = diff(J,w)
dJ_dw

3*w**2

In [19]:
dJ_dw.subs([(w,2)])   # derivative at the point w=2

12

Comparing this with the arithmetic calculation

In [20]:
J = (2)**3
J_epsilon = (2+0.001)**3
k = (J_epsilon - J)/0.001
print(f"J = {J}, J_epsilon = {J_epsilon}, dJ_dw ~= k = {k} ")

J = 8, J_epsilon = 8.012006000999998, dJ_dw ~= k = 12.006000999997823 


𝐽=1𝑤

In [21]:
J, w = symbols('J, w')

In [22]:
J= 1/w
J

1/w

In [23]:
dJ_dw = diff(J,w)
dJ_dw

-1/w**2

In [24]:
dJ_dw.subs([(w,2)])

-1/4

Comparing this with the arithmetic calculation

In [25]:
J = 1/2
J_epsilon = 1/(2+0.001)
k = (J_epsilon - J)/0.001
print(f"J = {J}, J_epsilon = {J_epsilon}, dJ_dw ~= k = {k} ")

J = 0.5, J_epsilon = 0.49975012493753124, dJ_dw ~= k = -0.2498750624687629 


## Final Outcome

understood that a derivative describes the change in the output of a function that is a result of a small change in an input to that function. We also can use SymPy in python to find the symbolic derivative of functions.